src/v/cluster/tm_stm.cc

```diff
@@ -102,115 +133,161 @@ tm_stm::update_tx(tm_transaction tx, model::term_id term) {
+    auto tx = tx_opt.value();
+    if (
+      tx.status != tm_transaction::tx_status::ongoing
+      && tx.status != tm_transaction::tx_status::preparing) {
+        co_return tm_stm::op_status::conflict;
+    }
+    tx.status = cluster::tm_transaction::tx_status::killed;
+    tx.last_update_ts = clock_type::now();
+    co_return co_await update_tx(tx, expected_term);
 }
 
-ss::future<checked<tm_transaction, tm_stm::op_status>>
-tm_stm::mark_tx_ready(kafka::transactional_id tx_id) {
-    return mark_tx_ready(tx_id, _insync_term);
+ss::future<checked<tm_transaction, tm_stm::op_status>> tm_stm::reset_tx_ready(
+  model::term_id expected_term, kafka::transactional_id tx_id) {
+    return reset_tx_ready(expected_term, tx_id, expected_term);
 }
 
-ss::future<checked<tm_transaction, tm_stm::op_status>>
-tm_stm::mark_tx_ready(kafka::transactional_id tx_id, model::term_id term) {
-    auto ptx = _tx_table.find(tx_id);
-    if (ptx == _tx_table.end()) {
+ss::future<checked<tm_transaction, tm_stm::op_status>> tm_stm::reset_tx_ready(
+  model::term_id expected_term,
+  kafka::transactional_id tx_id,
+  model::term_id term) {
+    auto tx_opt = get_tx(tx_id);
+    if (!tx_opt.has_value()) {
         co_return tm_stm::op_status::not_found;
     }
-    tm_transaction tx = ptx->second;
+    tm_transaction tx = tx_opt.value();
     tx.status = tm_transaction::tx_status::ready;
     tx.partitions.clear();
     tx.groups.clear();
     tx.etag = term;
     tx.last_update_ts = clock_type::now();
-    co_return co_await update_tx(tx, _insync_term);
+    co_return co_await update_tx(tx, expected_term);
 }
 
 checked<tm_transaction, tm_stm::op_status>
 tm_stm::mark_tx_ongoing(kafka::transactional_id tx_id) {
-    auto ptx = _tx_table.find(tx_id);
-    if (ptx == _tx_table.end()) {
+    auto tx_opt = get_tx(tx_id);
+    if (!tx_opt.has_value()) {
         return tm_stm::op_status::not_found;
     }
-    ptx->second.status = tm_transaction::tx_status::ongoing;
-    ptx->second.tx_seq += 1;
-    ptx->second.partitions.clear();
-    ptx->second.groups.clear();
-    ptx->second.last_update_ts = clock_type::now();
-    ptx->second.etag = _insync_term;
-    return ptx->second;
+    tm_transaction tx = tx_opt.value();
+    tx.status = tm_transaction::tx_status::ongoing;
+    tx.tx_seq += 1;
+    tx.partitions.clear();
+    tx.groups.clear();
+    tx.last_update_ts = clock_type::now();
+    _mem_txes[tx_id] = tx;
+    return tx;
+}
+
+checked<tm_transaction, tm_stm::op_status>
+tm_stm::reset_tx_ongoing(kafka::transactional_id tx_id, model::term_id term) {
+    auto tx_opt = get_tx(tx_id);
+    if (!tx_opt.has_value()) {
+        return tm_stm::op_status::not_found;
+    }
+    tm_transaction tx = tx_opt.value();
+    tx.status = tm_transaction::tx_status::ongoing;
+    tx.tx_seq += 1;
+    tx.partitions.clear();
+    tx.groups.clear();
+    tx.last_update_ts = clock_type::now();
+    tx.etag = term;
+    _mem_txes[tx_id] = tx;
+    return tx;
 }
```

Comment:
nit: looks like the tx could be reset in-place.

Reply:
get_tx doesn't necessary return the struct from _mem_txes so we can't do it in-place

Reply:
got it

> doesn't necessary return

but it sounds like in some cases it can be done in place. keep an eye open for optimization opportunities even if we have to make the interfaces a little more transparent. all the copying really does add up over time.